src/scanspec/core.py

```diff
@@ -174,13 +200,41 @@ def _rebuild(cls_or_func: Callable[..., T]) -> None:
             if issubclass(cls_or_func, BaseModel):
                 cls_or_func.model_rebuild(force=True)
 
-    def schema(self, handler: GetCoreSchemaHandler) -> CoreSchema:
+    def schema(self, actual_type: type, handler: GetCoreSchemaHandler) -> CoreSchema:
         return tagged_union_schema(
-            _make_schema(tuple(self._subclasses), handler),
+            _make_schema(tuple(self._specify_generics(sub, actual_type) for sub in self._subclasses), handler),
             discriminator=self._discriminator,
             ref=self._base_class.__name__,
         )
 
+    def _support_subclass(self, subcls: type) -> bool:
+        if subcls == self._base_class:
+            return True
+        sub_params = _parameters(subcls)
+        if len(self._generics) != len(sub_params):
+            return False
+        if not all(_compatible_types(l, r) for l, r in zip(self._generics, sub_params)):
+            return False
+        if any(not self._support_subclass(get_origin(base) or base) for base in get_original_bases(subcls)):
+            return False
+        return True
+
+    def _specify_generics(self, subcls: type, actual_type: type) -> type | GenericAlias:
+        args = get_args(actual_type)
+        if args:
+            return GenericAlias(subcls, args)
+        return subcls
+
+def _parameters(possibly_generic: type) -> tuple[Any,...]:
+    return getattr(possibly_generic, '__parameters__', ())
+
+def _compatible_types(left: TypeVar, right: TypeVar) -> bool:
+    return (
+            left.__bound__ == right.__bound__
+            and left.__constraints__ == right.__constraints__
+            and left.__covariant__ == right.__covariant__
+            and left.__contravariant__ == right.__contravariant__
+            )
```

Comment:
Could left.__ be a subset of right.__?

```
class Super:
class Sub(Super):

@discriminated...
class Parent(Generic[Super]):
class Child(Parent[Sub], Generic[Sub])
```

e.g. for class ReadableMovable(Movable)?

Reply:
You can't pass concrete types to `Generic` as in this example but yes, this is definitely restricting the supported cases beyond what is possible in theory. It turns out to be difficult to determine what should be passed without re-implementing an entire type-checker. There are a couple of examples in #156 and a few in a previous version of this commit where I tried to support a wider range of options but until we have a use case for them I'd prefer a well defined restriction instead of an array of unsupported edge-cases.